### PR TITLE
fix: do not allow create pick list if pick list qty exceed sales order qty

### PR DESCRIPTION
**Pull request by vishdha**
TASK ID: https://bloomstack.com/desk#Form/Task/TASK-2020-01421
Peek 2020-09-15 16-41

### CODE REVIEW

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 1]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
 		self.delivery_date = min(order_delivery_dates)
 
 	def validate_stock_qty(self):
+		"""User should not allowed to create pick list if sales order item qty exceed"""
+		pick_list_items = frappe.get_all("Pick List Item", fields=["*"])
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: You don't seem to be using this

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 2]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
...
+				item_sales_order_qty = frappe.db.get_value("Sales Order Item", item.get("sales_order_item"), "qty")
+				used_qty = frappe.get_all("Pick List Item", filters={"sales_order_item":item.get("sales_order_item"), "docstatus":1}, fields=['sum(qty) as used_qty'])
+
+				for val in used_qty:
+					used_qty = val.used_qty
+
+				if used_qty:
+					pick_list_qty = item_sales_order_qty - used_qty
+					if pick_list_qty > 0 and item.qty > pick_list_qty:
+						frappe.throw(_("Row #{0}: Pick List thershold qty exceed for {1}'s . you are allowed to create pick list of {2} qty").format(item.idx, frappe.bold(item.item_name), frappe.bold(pick_list_qty)))
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]:
```suggestion
						frappe.throw(_("Row #{0}: Picked quantity ({1}) for {2} cannot exceed ordered qty ({3})").format(item.idx, frappe.bold(pick_list_qty), frappe.bold(item.item_name), frappe.bold(item_sales_order_qty)))
```

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 3]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
...
 				order_qty = frappe.db.get_value("Sales Order Item", item.get("sales_order_item"), "qty")
 				if item.qty > order_qty:
-					frappe.throw(_("Row #{0}: {1}'s quantity ({2}) should be less than or equal to the ordered quantity ({3})").format(
+					frappe.throw(_("Row #{0}: {1}'s quantity {2} should be less than or equal to the ordered quantity {3}").format(
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: Keep the parentheses, they look off in the normal sentence structure

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 4]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
 		self.delivery_date = min(order_delivery_dates)
 
 	def validate_stock_qty(self):
+		"""User should not allowed to create pick list if sales order item qty exceed"""
+		pick_list_items = frappe.get_all("Pick List Item", fields=["*"])
 		for item in self.locations:
 			if item.get("sales_order_item"):
+				pick_list_qty = 0
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: You're initializing this again later, you don't need it here

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 5]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
...
+				item_sales_order_qty = frappe.db.get_value("Sales Order Item", item.get("sales_order_item"), "qty")
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: If there is no sales order in the row, this'll fail at row 63

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 6]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
...
+				used_qty = frappe.get_all("Pick List Item", filters={"sales_order_item":item.get("sales_order_item"), "docstatus":1}, fields=['sum(qty) as used_qty'])
+
+				for val in used_qty:
+					used_qty = val.used_qty
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]:
```suggestion
				if used_qty:
					used_qty = used_qty[0].used_qty
```

File: erpnext/stock/doctype/pick_list/pick_list.py [hunk 7]
```diff
@@ -47,11 +48,24 @@ def validate_delivery_date(self):
 		self.delivery_date = min(order_delivery_dates)
 
 	def validate_stock_qty(self):
+		"""User should not allowed to create pick list if sales order item qty exceed"""
+		pick_list_items = frappe.get_all("Pick List Item", fields=["*"])
 		for item in self.locations:
 			if item.get("sales_order_item"):
+				pick_list_qty = 0
+				item_sales_order_qty = frappe.db.get_value("Sales Order Item", item.get("sales_order_item"), "qty")
+				used_qty = frappe.get_all("Pick List Item", filters={"sales_order_item":item.get("sales_order_item"), "docstatus":1}, fields=['sum(qty) as used_qty'])
```
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: Call this `prev_picked_qty`

File: erpnext/stock/doctype/pick_list/pick_list.py (hunk 5 above)
Comment (Alchez) [inline erpnext/stock/doctype/pick_list/pick_list.py]: Call this `ordered_item_qty`

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/pick_list/pick_list.py (hunk 5 above)
Comment (vishdha) [reply erpnext/stock/doctype/pick_list/pick_list.py]: line 54 has check for sales order item

**Review by neilLasrado — APPROVED** [review]

**Pull request merged**